### PR TITLE
update Go version; add websocket / correct endpoints to 'src gateway benchmark'

### PR DESCRIPTION
**Pull request by slimsag**
Continues development of #1124

* Updates Go version used in this repository (I had to do this to develop, see https://sourcegraph.slack.com/archives/C07KZF47K/p1731627487087909)
* Checks the status code returned by HTTP endpoints.
* Updates Sourcegraph HTTP endpoints to `/.api/gateway` and `/.api/gateway/websocket`.
* Updates gateway HTTP endpoints to `/v2` and `/v2/websocket`
* Adds websocket client.
* Respects `-gateway` parameter and `SRC_ENDPOINT`/`SRC_ACCESS_TOKEN`

This should be sufficient for testing against the actual production endpoints once they are added / working / live in prod.

### Test plan

Not tested yet since endpoints are not live yet. Command not in use yet, so no risk.

### CODE REVIEW

File: cmd/src/gateway_benchmark.go
Comment (vdavid) [inline cmd/src/gateway_benchmark.go:65]: I think this would help clarify that `any`.

```suggestion
			endpoints                              = map[string]any{} // Values: URL `string`s or `*websocket.Conn`s
```

File: cmd/src/gateway_benchmark.go
Comment (vdavid) [inline cmd/src/gateway_benchmark.go:68]: Ah, the `http` → `ws` replacement is nice and elegant to keep that optional `s`. Self-explanatory once I saw it, but I missed it in my original version. Thanks! 👏

File: cmd/src/gateway_benchmark.go
Comment (vdavid) [inline cmd/src/gateway_benchmark.go:74]: Nit, maybe this would be more precise? We output these names, so might be worth adjusting.

```suggestion
			endpoints["ws(s): gateway"] = gatewayWebsocket
			endpoints["http(s): gateway"] = fmt.Sprint(*gatewayEndpoint, "/v2")
```
(same for sourcegraph endpoints below)

File: go.sum
Comment (vdavid) [inline go.sum:1]: I think this update messed up something, see the check annotations below. Also, would be nice to mention in the PR description that you updated packages, and state the reason (incl. whether it was related or unrelated to the main purpose of the PR).

File: cmd/src/gateway_benchmark.go
Comment (vdavid) [inline cmd/src/gateway_benchmark.go:170]: Ah, thanks! This was IntelliJ's autogenerated code. Will watch out for this in the future.

File: cmd/src/gateway_benchmark.go
Comment (vdavid) [inline cmd/src/gateway_benchmark.go:184]: Maybe check the body as well? I think it's a great call that you checked for response body == "pong" for WebSockets. I think we should do the same here. I'll adjust the endpoints to respond with "pong"

File: cmd/src/gateway_benchmark.go
Comment (vdavid) [inline cmd/src/gateway_benchmark.go:189]: Naming: I'd have called this `benchmarkEndpointWebsocket` with a non-capital "s". I looked up the Go docs and your version seems more idiomatic with Go.
If we agree that "WebSocket" is the correct capitalization for entity names in Go then please also rename `gatewayWebsocket` and `sourcegraphWebsocket` accordingly. (I'd do it to fix my mistake but I don't want to mess or conflict with your PR.)

**Review by vdavid — APPROVED** [review]
Approved to unblock and because it's generally 👌. Also left a few suggestions to improve it further before merging.

File: cmd/src/gateway_benchmark.go
Comment (vdavid) [reply cmd/src/gateway_benchmark.go:189]: I've changed it to "WebSocket" everywhere in https://github.com/sourcegraph/sourcegraph/pull/1526

File: cmd/src/gateway_benchmark.go
Comment (vdavid) [inline cmd/src/gateway_benchmark.go:201]: I've changed my other PR https://github.com/sourcegraph/sourcegraph/pull/1526 to respond with "pong" on all three endpoints.

File: cmd/src/gateway_benchmark.go
Comment (slimsag) [reply cmd/src/gateway_benchmark.go:189]: wikipedia says WebSocket so yea let's go with that capitalization

File: cmd/src/gateway_benchmark.go
Comment (slimsag) [reply cmd/src/gateway_benchmark.go:201]: Fantastic!

File: cmd/src/gateway_benchmark.go
Comment (vdavid) [reply cmd/src/gateway_benchmark.go:189]: I was aware of that, I just wasn't aware how to translate MixedCase entity names into Go entity names. But yeah, all clear, WebSocket it is!

File: go.sum
Comment (slimsag) [reply go.sum:1]: Yeah, so I think we're going to run into some challenges here with getting this PR merged:

* Currently, I can't develop src-cli at all without modifying the go.mod file to have a replace directive pointing to the private monorepo: https://sourcegraph.slack.com/archives/C07KZF47K/p1731627487087909
* With the replace directive, the Go version of this repository must be updated.. which requires updating the Go packages.
* I think(?) this repository now uses some Bazel shenanigans and that will need to be udpated to reflect these two bullet points - or else I'm guessing this PR will fail CI.

In short, I don't think we can currently merge changes to src-cli repository at all :/ will need to sort this out seperately before we can merge this PR.

Comment (vdavid) [reply go.sum:1]: Crap, what an unexpected skeleton falling out of that closet. Thanks for handling it!

**Review by camdencheek — APPROVED** [review]
Didn't review the changes to `src gateway benchmark` closely, but stamping on behalf of the code search team for the version bumps. Thanks!

**Pull request merged**
